### PR TITLE
[REF] OAuth - Move some API code to the BAO for better reusability

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Moves some code out of the API into the BAO so we get more use out of it.

Before
----------------------------------------
Lots of overridden API actions, which is not standard practice.

After
----------------------------------------
Instead of overriding API actions, use the standard actions and put the other logic in the correct place in the BAOs so it gets picked up automatically.

Comments
------------------------
I haven't used this extension or tested any of this code, this is just what I think the structure ought to look like.

### CODE REVIEW

**Comment by colemanw** [comment]
FYI @highfalutin this refactors your api code to move logic around for better reusability. Hopefully I've kept your functionality the same during the refactor.

**Comment by demeritcowboy** [comment]
Just looking at a few they are all `Error: Call to undefined function getLoggedInContactID()`

[Civi\OAuth\AuthCodeFlowTest.testContactToken_AnonymousUser_LinkExistingContactRecord](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_AnonymousUser_LinkExistingContactRecord/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_AnonymousUser_SetNullContactId](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_AnonymousUser_SetNullContactId/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_AnonymousUser_CreateContact](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_AnonymousUser_CreateContact/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_LoggedInUser_Default](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_LoggedInUser_Default/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_LoggedInUser_LinkOtherContact](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_LoggedInUser_LinkOtherContact/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_LoggedInUser_CreateContact](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_LoggedInUser_CreateContact/)
    [Civi\OAuth\AuthCodeFlowTest.testContactToken_LoggedInUser_SetNullContactId](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/Civi_OAuth_AuthCodeFlowTest/testContactToken_LoggedInUser_SetNullContactId/)
    [api_v4_OAuthContactTokenTest.testCreate](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testCreate/)
    [api_v4_OAuthContactTokenTest.testRead](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testRead/)
    [api_v4_OAuthContactTokenTest.testUpdate](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testUpdate/)
    [api_v4_OAuthContactTokenTest.testDelete](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testDelete/)
    [api_v4_OAuthContactTokenTest.testGetByScope](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testGetByScope/)
    [api_v4_OAuthContactTokenTest.testPostHook](https://test.civicrm.org/job/CiviCRM-Core-PR/49575/testReport/junit/(root)/api_v4_OAuthContactTokenTest/testPostHook/)

File: ext/oauth-client/CRM/OAuth/BAO/OAuthContactToken.php [hunk 1]
```diff
@@ -2,4 +2,121 @@
...
+   * @return bool
+   * @see CRM_Core_DAO::checkAccess
+   */
+  public static function _checkAccess(string $entityName, string $action, array $record, $userId): bool {
+    try {
+      $record['check_permissions'] = TRUE;
+      self::fillAndValidate($record, $userId);
+      return TRUE;
+    }
+    catch (\Civi\API\Exception\UnauthorizedException) {
```
Comment (demeritcowboy) [inline ext/oauth-client/CRM/OAuth/BAO/OAuthContactToken.php:42]: It doesn't seem to like this. It wants a variable like `\Civi\API\Exception\UnauthorizedException $ue`. The error is `ParseError: syntax error, unexpected ')', expecting '|' or variable (T_VARIABLE)`. Not sure if this is php-version specific but I was using 7.4.

**Comment by demeritcowboy** [comment]
I did two simple tests. Except for the above comment:
* The basic test described in the `r-run` at https://github.com/civicrm/civicrm-core/pull/20189#issuecomment-844534735 works.
* My existing test setup with a mail account and using fetch_activities (which is a systoken not a contact token, but there's some other refactoring here), still works.

Have not done code review.

**Comment by demeritcowboy** [comment]
Also FYI when testing make sure your test site supports https and CIVICRM_BASEURL is https. And also Google has made the oauth prompt about unverified apps even scarier than before, but will still let you continue.

File: ext/oauth-client/CRM/OAuth/BAO/OAuthContactToken.php (hunk 1 above)
Comment (colemanw) [reply ext/oauth-client/CRM/OAuth/BAO/OAuthContactToken.php]: Good "catch" @demeritcowboy. Fixed, thanks.

**Comment by demeritcowboy** [comment]
This has a merge conflict: ext/oauth-client/Civi/Api4/Action/OAuthContactToken/Create.php

**Comment by colemanw** [comment]
@demeritcowboy I've rebased to fix the MC

**Comment by highfalutin** [comment]
Hi @colemanw, getting a chance to look at this now. When you say this "Moves some code out of the API into the BAO so we get more use out of it", what uses are you thinking of? Search Kit?

I support the code being refactored into a normal form as long as everything still works. I'm just curious about what the payoff will be.

**Comment by highfalutin** [comment]
I did my own `r-run`/`r-test`, with good results.

I installed the code from this PR and ran the automated tests for a [custom extension](https://github.com/lemniscus/osmf-verify-contributor) that depends on `oauth-client`. These tests don't directly test the permissions mechanisms, but the fact that they all passed is good. It tells me this PR doesn't break anything, as far as I can tell.

I also verified that the `oauth-client` tests DO test the permissions mechanisms, and they are all passing.

**Comment by colemanw** [comment]
Merging based on review

**Comment by demeritcowboy** [comment]
Thanks @highfalutin @colemanw

**Pull request merged**